### PR TITLE
Fixes #29371 - Update settings.py for pulpcore3.2

### PR DESCRIPTION
**Pull request by sjha4**
We are working on getting pulpcore3.2 built and released. This is required to support the changes required to pulp/settings 

@wbclark : This is the the only place I identified a need for changes to support pulpcore 3.2 with this module. Is there anything else that you can think of that needs to be done?

### CODE REVIEW

File: templates/settings.py.erb [hunk 1]
```diff
@@ -27,3 +27,11 @@ WORKING_DIRECTORY = "<%= scope['pulpcore::cache_dir'] %>"
...
 REMOTE_USER_ENVIRON_NAME = '<%= scope['pulpcore::remote_user_environ_name'] %>'
 <% end -%>
+AUTHENTICATION_BACKENDS = ['pulpcore.app.authentication.PulpNoCreateRemoteUserBackend']
+
+REST_FRAMEWORK__DEFAULT_AUTHENTICATION_CLASSES = (
+    'rest_framework.authentication.SessionAuthentication',
+    'pulpcore.app.authentication.PulpRemoteUserAuthentication'
+)
+
+ALLOWED_IMPORT_PATHS = ['/var/www/test_repos']
```
Comment (ehelms) [inline templates/settings.py.erb]: What is this line handling?

Comment (jlsherrill) [reply templates/settings.py.erb]: Pulp 3 added a restriction on what paths you can sync from using the 'file://' 'protocol'.    So this is needed to allow our tests to run, however @sjha4 i think we probably want to set this to a more generic path for users and probably one that would allow python to optimize the copy (which later versions of python do).  I would recommend /var/lib/pulp/sync_imports/  or something like that.  And we'll need to update our tests to use that

Comment (sjha4) [reply templates/settings.py.erb]: Yes..this is to handle our tests only for now..Let me try modifying tests with this /var/lib/pulp/sync_imports/ and check if I run into any issues.. :+1:

Comment (ehelms) [reply templates/settings.py.erb]: If only for testing, a comment in this case would be useful as that is information I'm sure will be lost on us down the road.

Comment (sjha4) [reply templates/settings.py.erb]: Added the comment and updated katello tests to use path '/var/lib/pulp/sync_imports/'

Comment (jlsherrill) [reply templates/settings.py.erb]: @ehelms its not *only* for tests, pulp will ONLY sync from file:// repos in the directories we specify, this is a change from pulp2.  Our tests that do this need to use this directory as well.

**Comment by jlsherrill** [comment]
ACK from me, but i think this needs to wait to be merged until https://github.com/Katello/katello/pull/8617 is ready

File: templates/settings.py.erb [hunk 2]
```diff
@@ -27,3 +27,12 @@ WORKING_DIRECTORY = "<%= scope['pulpcore::cache_dir'] %>"
...
 <% end -%>
+AUTHENTICATION_BACKENDS = ['pulpcore.app.authentication.PulpNoCreateRemoteUserBackend']
+
+REST_FRAMEWORK__DEFAULT_AUTHENTICATION_CLASSES = (
+    'rest_framework.authentication.SessionAuthentication',
+    'pulpcore.app.authentication.PulpRemoteUserAuthentication'
+)
+
+<%# This setting whitelists paths that can be used for repository sync with file protocol. Katello uses the path /var/lib/pulp/sync_imports/ to run tests%>
+ALLOWED_IMPORT_PATHS = ['/var/lib/pulp/sync_imports/']
```
Comment (ekohl) [inline templates/settings.py.erb]: It's a good practice to always use variables for paths. Even if you don't want to change it, this allows reusing this value if Katello needs it. It also allows setting it to an empty array in production for additional security.

File: templates/settings.py.erb [hunk 3]
```diff
@@ -27,3 +27,12 @@ WORKING_DIRECTORY = "<%= scope['pulpcore::cache_dir'] %>"
...
 REMOTE_USER_ENVIRON_NAME = '<%= scope['pulpcore::remote_user_environ_name'] %>'
 <% end -%>
+AUTHENTICATION_BACKENDS = ['pulpcore.app.authentication.PulpNoCreateRemoteUserBackend']
+
+REST_FRAMEWORK__DEFAULT_AUTHENTICATION_CLASSES = (
+    'rest_framework.authentication.SessionAuthentication',
+    'pulpcore.app.authentication.PulpRemoteUserAuthentication'
+)
+
+<%# This setting whitelists paths that can be used for repository sync with file protocol. Katello uses the path /var/lib/pulp/sync_imports/ to run tests%>
```
Comment (ekohl) [inline templates/settings.py.erb]: IMHO it's fine to place this as a Python comment rather than an ERB comment. If you keep it as an ERB comment, please strip the trailing whitespace (`-%>` instead of `%>`).

File: manifests/init.pp [hunk 4]
```diff
@@ -110,6 +113,7 @@
   String $django_secret_key = extlib::cache_data('pulpcore_cache_data', 'secret_key', extlib::random_password(32)),
   Integer[0] $redis_db = 8,
   Stdlib::Fqdn $servername = $facts['fqdn'],
+  Stdlib::Absolutepath $allowed_import_path = '/var/lib/pulp/sync_imports',
   Optional[String] $remote_user_environ_name = undef,
```
Comment (sjha4) [inline manifests/init.pp]: @ekohl: I was trying to look it up but how to do I define this to be an array of absolute paths?

Comment (sjha4) [reply manifests/init.pp]: Looks like I can do this https://github.com/theforeman/puppet-pulpcore/blob/c27d67a6c9705b99c942729dd539698e734f6bd3/manifests/admin.pp#L28

Comment (ekohl) [reply manifests/init.pp]: Exactly. I love it when people can answer their own questions :)

File: templates/settings.py.erb [hunk 5]
```diff
@@ -27,3 +27,12 @@ WORKING_DIRECTORY = "<%= scope['pulpcore::cache_dir'] %>"
...
 <% end -%>
+AUTHENTICATION_BACKENDS = ['pulpcore.app.authentication.PulpNoCreateRemoteUserBackend']
+
+REST_FRAMEWORK__DEFAULT_AUTHENTICATION_CLASSES = (
+    'rest_framework.authentication.SessionAuthentication',
+    'pulpcore.app.authentication.PulpRemoteUserAuthentication'
+)
+
+<%# This setting whitelists paths that can be used for repository sync with file protocol. Katello uses the path /var/lib/pulp/sync_imports/ to run tests -%>
+ALLOWED_IMPORT_PATHS = "<%= scope['pulpcore::allowed_import_path'] %>"
```
Comment (ekohl) [inline templates/settings.py.erb]: I think this needs to be a loop to construct an array.

**Comment by sjha4** [comment]
@ekohl : Added some tests and looping over param to get the correct array.

File: spec/classes/pulpcore_spec.rb [hunk 6]
```diff
@@ -120,6 +120,30 @@
...
+      context 'with allowed import paths' do
+        let :params do
+          {
+            allowed_import_path: ['/test/path', '/test/path2'],
+          }
+        end
+
+        it do
+          is_expected.to compile.with_all_deps
+	  is_expected.to contain_concat__fragment('base')
```
Comment (ekohl) [inline spec/classes/pulpcore_spec.rb]: Nit: looks like there's a tab here

File: templates/settings.py.erb [hunk 7]
```diff
@@ -27,3 +27,17 @@ WORKING_DIRECTORY = "<%= scope['pulpcore::cache_dir'] %>"
...
 <% end -%>
+AUTHENTICATION_BACKENDS = ['pulpcore.app.authentication.PulpNoCreateRemoteUserBackend']
+
+REST_FRAMEWORK__DEFAULT_AUTHENTICATION_CLASSES = (
+    'rest_framework.authentication.SessionAuthentication',
+    'pulpcore.app.authentication.PulpRemoteUserAuthentication'
+)
+
+<%# This setting whitelists paths that can be used for repository sync with file protocol. Katello uses the path /var/lib/pulp/sync_imports/ to run tests -%>
+<% allowed_paths = [] -%>
```
Comment (ekohl) [inline templates/settings.py.erb]: Why is this needed? If anything you can use `ALLOWED_IMPORT_PATHS = <%= Array.new(scope['pulpcore::allowed_import_path']) %>` but I think it should not be needed since it's already an array.

Comment (sjha4) [reply templates/settings.py.erb]: That is correct..Updated..This works `scope['pulpcore::allowed_import_path']` .. :+1:

**Comment by jlsherrill** [comment]
These two settings:

AUTHENTICATION_BACKENDS
REST_FRAMEWORK__DEFAULT_AUTHENTICATION_CLASSES 

are needed for katello's cert-auth to work, and may cause previous versions of pulp to not function properly, although there were so many bugs around it within pulp, it may just ignore them alltogether.  I can give it a test real quick

**Comment by jlsherrill** [comment]
Testing revealed that it does in fact break pulpcore 3.0 because 'PulpNoCreateRemoteUserBackend' is not defined

**Comment by ekohl** [comment]
I don't mind breaking it, but I want to clearly know that we did so we don't accidentally cherry pick it. It'd also be good to have this in the README.

**Comment by sjha4** [comment]
@jlsherrill , @ekohl : Updated Readme to add some information on the new settings detailing these are meant to be used in 3.2
